core/src/apps/common/request_pin.py

```diff
@@ -80,33 +80,41 @@ class PinInvalid(Exception):
 async def request_pin_and_sd_salt(
     ctx: wire.Context, prompt: str = "Enter your PIN", allow_cancel: bool = True
 ) -> Tuple[str, Optional[bytearray]]:
-    salt = await request_sd_salt(ctx)
-
     if config.has_pin():
         pin = await request_pin_ack(ctx, prompt, config.get_pin_rem(), allow_cancel)
+        config.ensure_not_wipe_code(pin_to_int(pin))
     else:
         pin = ""
 
+    salt = await request_sd_salt(ctx)
+
     return pin, salt
 
 
 async def verify_user_pin(
     prompt: str = "Enter your PIN", allow_cancel: bool = True, retry: bool = True
 ) -> None:
+    if config.has_pin():
+        pin = await request_pin(prompt, config.get_pin_rem(), allow_cancel)
+        config.ensure_not_wipe_code(pin_to_int(pin))
+    else:
+        pin = ""
+
     try:
         salt = await request_sd_salt()
     except SdCardUnavailable:
         raise PinCancelled
-
-    if not config.has_pin() and not config.check_pin(pin_to_int(""), salt):
+    if config.unlock(pin_to_int(pin), salt):
+        return
+    elif not config.has_pin():
         raise RuntimeError
 
     while retry:
```

Comment:
ISTM previously this function would always show the PIN dialog. Now it unlocks if PIN is not enabled.

That seems like a reasonable behavior from a function "verify user pin". But the other usage of this function is in FIDO `verify_user`, where no check for PIN enabled was done.
Was that a bug, or did the check happen somewhere else?
Or is it now a bug that the function will not ask for PIN if it is not enabled?

Reply:
You are correct, I should have mentioned that.

In FIDO a disabled PIN would be caught here:
https://github.com/trezor/trezor-firmware/blob/e51ed5bf05e9f4627b228ce6874341ed82ff4222/core/src/apps/webauthn/fido2.py#L1425
https://github.com/trezor/trezor-firmware/blob/e51ed5bf05e9f4627b228ce6874341ed82ff4222/core/src/apps/webauthn/fido2.py#L1578
So it couldn't manifest as a bug, because the calling code ensured the precondition. Nevertheless I think it was a bug in the sense that this precondition was not tested or documented. I prefer this behavior which checks for empty PIN and unlocks / passes automatically, since it seems generally more useful.

On a side note, there was another bug in the code previously, if `retry=False`, but it was never used with this parameter.